linera-core/src/client/mod.rs

```diff
@@ -2458,36 +2369,22 @@ where
     async fn process_pending_block_without_prepare(
         &self,
     ) -> Result<ClientOutcome<Option<ConfirmedBlockCertificate>>, ChainClientError> {
-        let identity = self.identity().await?;
-        let mut info = self.chain_info_with_manager_values().await?;
-        // If the current round has timed out, we request a timeout certificate and retry in
-        // the next round.
-        if let Some(round_timeout) = info.manager.round_timeout {
-            if round_timeout <= self.storage_client().clock().current_time() {
-                self.request_leader_timeout().await?;
-                info = self.chain_info_with_manager_values().await?;
-            }
-        }
-        self.update_from_info(&info);
-        let manager = &info.manager;
+        let info = self.request_leader_timeout_if_needed().await?;
 
+        let identity = self.identity().await?;
         // If there is a validated block in the current round, finalize it.
-        if let Some(certificate) = &manager.requested_locked {
-            if certificate.round == manager.current_round {
+        if let Some(certificate) = &info.manager.requested_locked {
+            if certificate.round == info.manager.current_round {
```

Comment:
Personal preference, so you can ignore if disagree, but I'd extract lines 2378-2394 to `fn finalize_current_round`. 

I'd even consider going as far as move the `if let Some(certificate) = &info.manager.requested_locked ...` there already. So the caller here would call `self.finalize_current_round(&info).await?;` which might not do anything but from the high level this is the correct order of prioririties.

Reply:
Or maybe `finalize_locked` 🤷

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/2938/commits/c89e9bc66da007b8bda9d331e98c9041e79c8c33.